### PR TITLE
Allow for null literals

### PR DESCRIPTION
**Pull request by chadhietala**

### CODE REVIEW

**Comment by chancancode** [comment]
Does other kind of literals (true false strings numbers etc) have this problem? Also, is undefined actually a thing we support ?_?

File: packages/glimmer-wire-format/index.ts [hunk 1]
```diff
@@ -78,7 +78,10 @@ export namespace Expressions {
   export const isHasBlockParams = is<HasBlockParams>('hasBlockParams');
 
   export function isValue(value: any): value is Value {
-    return value !== null && typeof value !== 'object';
+    if (value === null) {
+      return true;
```
Comment (rwjblue) [inline packages/glimmer-wire-format/index.ts]: This confuses me. If `value` is `null` the `isValue` method is supposed to return true? Seems to me like null/undefined are **not** values?

**Comment by rwjblue** [comment]
Can we also have a test for when s bound value resolves to null/undefined (vs just using the null/undefined literal in the template).

**Comment by chancancode** [comment]
@rwjblue I _think_ we already have tests for those, and they are pretty
unrelated to this at this level (syntax). But if we somehow don't have them
then yeah we should add some.

**Comment by chancancode** [comment]
(Those tests would be in the updating tests file)

File: packages/glimmer-wire-format/index.ts (hunk 1 above)
Comment (chadhietala) [inline packages/glimmer-wire-format/index.ts]: Changed this to `isPrimitiveValue`

File: packages/glimmer-runtime/lib/syntax/expressions.ts [hunk 2]
```diff
@@ -24,11 +24,11 @@ const {
...
   isUnknown,
-  isValue
+  isPrimitiveValue
 } = SerializedExpressions;
 
 export default function(sexp: SerializedExpression): any {
-  if (isValue(sexp)) {
+  if (isPrimitiveValue(sexp)) {
     return ValueSyntax.fromSpec(sexp);
   } else {
```
Comment (chancancode) [inline packages/glimmer-runtime/lib/syntax/expressions.ts]: Can you change this to use the same `if (...) return` syntax/style?

File: packages/glimmer-runtime/tests/initial-render-test.ts [hunk 3]
```diff
@@ -349,6 +349,46 @@ test("The compiler can handle simple helpers", function() {
   compilesTo('<div>{{testing title}}</div>', '<div>hello</div>', { title: 'hello' });
 });
 
+test("GH#13999 The compiler can handle simple helpers with inline null parameter", function() {
+  env.registerHelper('say-hello', function() {
+    return 'hello';
```
Comment (chancancode) [inline packages/glimmer-runtime/tests/initial-render-test.ts]: Can you assert that the argument is actually `null`?

Something like:

```
let value;

env.registerHelper('say-hello', function(params) {
  value = params[0];
  return hello;
});

strictEqual(value, ..., '...');
```

File: packages/glimmer-runtime/tests/initial-render-test.ts [hunk 4]
```diff
@@ -349,6 +349,46 @@ test("The compiler can handle simple helpers", function() {
...
+
+  compilesTo('<div>{{say-hello null}}</div>', '<div>hello</div>');
+});
+
+test("GH#13999 The compiler can handle simple helpers with inline undefined parameter", function() {
+  env.registerHelper('say-hello', function() {
+    return 'hello';
+  });
+
+  compilesTo('<div>{{say-hello undefined}}</div>', '<div>hello</div>');
```
Comment (chancancode) [inline packages/glimmer-runtime/tests/initial-render-test.ts]: I suspect that this is not something we actually support. The wire format is converted into json, which doesn't support `undefined` as a value. `undefined` feels very JavaScript-centric, whereas Handlebars try to stay a bit more host-language-neutural in general (there are implementations in other languages, and having `undefined` vs `null` in those environments would feel quite out-of-place).

File: packages/glimmer-runtime/tests/initial-render-test.ts [hunk 5]
```diff
@@ -349,6 +349,46 @@ test("The compiler can handle simple helpers", function() {
...
+});
+
+test("GH#13999 The compiler can handle simple helpers with inline undefined parameter", function() {
```
Comment (chancancode) [inline packages/glimmer-runtime/tests/initial-render-test.ts]: duplicated test?

File: packages/glimmer-runtime/tests/initial-render-test.ts [hunk 6]
```diff
@@ -349,6 +349,46 @@ test("The compiler can handle simple helpers", function() {
...
+});
+
+test("GH#13999 The compiler can handle components with undefined named arguments", function() {
+  env.registerHelper('say-hello', function() {
+    return 'hello';
+  });
+
+  compilesTo('<div>{{say-hello foo=undefined}}</div>', '<div>hello</div>');
+});
+
```
Comment (chancancode) [inline packages/glimmer-runtime/tests/initial-render-test.ts]: I can think of at least a few more cases to test – `{{null}}` ("append"/content position), `<div class="foo {{null}} bar" />` (concat) and `{{#if null}}` ("block" arguments).

**Comment by chancancode** [comment]
Looks good other than some minor things. This is pretty important, so we probably should test this in Ember side also (once this lands).

**Comment by chancancode** [comment]
Does TypeScript allow you to explicitly add ` | null` to https://github.com/tildeio/glimmer/blob/master/packages/glimmer-wire-format/index.ts#L43 without strict null-checking mode?

**Comment by chadhietala** [comment]
It is a tslint error to introduce nullable types in that union.

File: packages/glimmer-runtime/tests/initial-render-test.ts (hunk 6 above)
Comment (chadhietala) [inline packages/glimmer-runtime/tests/initial-render-test.ts]: Added these guys

**Comment by chancancode** [comment]
Maybe we should either configure tslint to ignore that line, or at least add `/* | null */;` for documentation

**Comment by chadhietala** [comment]
Just ignored the line from tslint.

**Comment by chancancode** [comment]
http://tryhandlebarsjs.com seems to suggest that Handlebars do support and differentiate between `undefined` and `null` (which I think is pretty strange/inappropriate):

tryhandlebarsjs

@wycats (@mmun?) can you clarify what we need to support here?

**Comment by chadhietala** [comment]
There is `UndefinedLiteral` and `NullLiteral`s

**Comment by chancancode** [comment]
I think there are a few options 1) maybe the language doesn't care and this is just a host extension, since we work in JS we support and differentiate both for convenience 2) the language requires us to support and differentiate both 3) we don't have to support undefined. 

Either way, it seems like you have found a bug: accepting undefined syntax but turning it into null doesn't seem acceptable; if we have to support it at all we would need to round trip it in the wire format somehow.

(That being said, we could land this fix and the tests and mark the undefined tests as bugs)

**Comment by chadhietala** [comment]
So from a back compat point of view `UndefinedLiteral`s resolve to `null` in HTMLBars, however from from an implementation point of view it is incorrect.

**Comment by chadhietala** [comment]
I have created https://github.com/emberjs/ember.js/issues/14016.
